File: backends/rails/lib/fixtures.rb
module Fixtures
  JOHN_EMAIL = "john@beatles.com"
  PAUL_EMAIL = "paul@beatles.com"
  MIKE_EMAIL = "mike@monsters.com"
  RINGO_EMAIL = "ringo@beatles.com"

  def self.load_fixture_data
    #########
    # Users #
    #########

    john = User.create(email: JOHN_EMAIL)
    paul = User.create(email: PAUL_EMAIL)
    admin = User.create(email: "admin@admin.com")
    mike = User.create(email: MIKE_EMAIL)
    sully = User.create(email: "sully@monsters.com")
    ringo = User.create(email: RINGO_EMAIL)
    randall = User.create(email: "randall@monsters.com")
    users = [
      john,
      paul,
      admin,
      mike,
      sully,
      ringo,
      randall,
    ]

    ########
    # Orgs #
    ########

    beatles = Org.create(
      name: "The Beatles",
      billing_address: "64 Penny Ln Liverpool, UK",
      base_repo_role: "reader",
    )
    monsters = Org.create(
      name: "Monsters Inc.",
      billing_address: "123 Scarers Rd Monstropolis, USA",
      base_repo_role: "reader",
    )
    orgs = [beatles, monsters]

    #########
    # Repos #
    #########

    abbey_road = Repo.create(name: "Abbey Road", org: beatles)
    paperwork = Repo.create(name: "Paperwork", org: monsters)
    repos = [abbey_road, paperwork]

    ##########
    # Issues #
    ##########

    too_much_critical_acclaim = Issue.create(title: "Too much critical acclaim", repo: abbey_road, creator: john)
    issues = [too_much_critical_acclaim]

    ##############
    # Repo roles #
    ##############

    # TODO

    RepoRole.create(user: john, repo: abbey_road, name: "reader")
    RepoRole.create(user: paul, repo: abbey_road, name: "reader")
    RepoRole.create(user: ringo, repo: abbey_road, name: "maintainer")
    RepoRole.create(user: mike, repo: paperwork, name: "reader")
    RepoRole.create(user: sully, repo: paperwork, name: "reader")

    # #############
    # # Org roles #
    # #############

    OrgRole.create(user: john, org: beatles, name: "owner")
    OrgRole.create(user: paul, org: beatles, name: "member")
    OrgRole.create(user: ringo, org: beatles, name: "member")
    OrgRole.create(user: mike, org: monsters, name: "owner")
    OrgRole.create(user: sully, org: monsters, name: "member")
    OrgRole.create(user: randall, org: monsters, name: "member")
  end
end

File: backends/flask-sqlalchemy/app/fixtures.py
from .models import Base, Issue, Org, Repo, User, OrgRole, RepoRole

john_email = "john@beatles.com"
paul_email = "paul@beatles.com"
mike_email = "mike@monsters.com"
ringo_email = "ringo@beatles.com"


def load_fixture_data(session):
    def org_role(user, org, name):
        role = OrgRole(user=user, org=org, name=name)
        session.add(role)

    def repo_role(user, repo, name):
        role = RepoRole(user=user, repo=repo, name=name)
        session.add(role)

    #########
    # Users #
    #########

    john = User(email=john_email)
    paul = User(email=paul_email)
    admin = User(email="admin@admin.com")
    mike = User(email=mike_email)
    sully = User(email="sully@monsters.com")
    ringo = User(email=ringo_email)
    randall = User(email="randall@monsters.com")
    users = [
        john,
        paul,
        admin,
        mike,
        sully,
        ringo,
        randall,
    ]
    for user in users:
        session.add(user)

    ########
    # Orgs #
    ########

    beatles = Org(
        name="The Beatles",
        billing_address="64 Penny Ln Liverpool, UK",
        base_repo_role="reader",
    )
    monsters = Org(
        name="Monsters Inc.",
        billing_address="123 Scarers Rd Monstropolis, USA",
        base_repo_role="reader",
    )
    orgs = [beatles, monsters]
    for org in orgs:
        session.add(org)

    #########
    # Repos #
    #########

    abby_road = Repo(name="Abbey Road", org=beatles)
    paperwork = Repo(name="Paperwork", org=monsters)
    repos = [abby_road, paperwork]
    for repo in repos:
        session.add(repo)

    ##########
    # Issues #
    ##########

    too_much_critical_acclaim = Issue(title="Too much critical acclaim", repo=abby_road)
    issues = [too_much_critical_acclaim]
    for issue in issues:
        session.add(issue)

    # https://github.com/osohq/oso/blob/70965f2277d7167c38d3641140e6e97dec78e3bf/languages/python/sqlalchemy-oso/tests/test_roles.py#L132-L133
    session.flush()

    ##############
    # Repo roles #
    ##############

    repo_role(john, abby_road, "reader")
    repo_role(paul, abby_road, "reader")
    repo_role(ringo, abby_road, "maintainer")
    repo_role(mike, paperwork, "reader")
    repo_role(sully, paperwork, "reader")

    #############
    # Org roles #
    #############

    org_role(john, beatles, "owner")
    org_role(paul, beatles, "member")
    org_role(ringo, beatles, "member")
    org_role(mike, monsters, "owner")
    org_role(sully, monsters, "member")
    org_role(randall, monsters, "member")

    session.flush()
    session.commit()
    session.close()
